editor/code/src/types.ts

```diff
@@ -0,0 +1,48 @@
+import {
+  TextDocumentIdentifier,
+  VersionedTextDocumentIdentifier,
+} from "vscode-languageclient";
+
+import { Position } from "vscode";
+
+export interface Pp {
+  hdr: string;
+  second?: Pp[] | string | number;
+  third?: Pp | number;
+}
+
+export enum BreakMode {
+  Horizontal = 0,
+  Vertical = 1,
+  SkipVertical = 2,
+}
+
+export interface PpState {
+  breakMode: BreakMode;
+  margin: string | null;
+}
+
+export interface EntryType {
+  [key: string]: any;
+}
+
+export interface Hyp {
+  names: string;
+  ty: EntryType;
+}
+
+export interface Goal {
+  ty: EntryType;
+  hyps: Hyp[];
+}
+
+export interface GoalRequest {
+  textDocument: TextDocumentIdentifier;
+  position: Position;
+}
+
+export interface GoalAnswer {
```

Comment:
Why is this a better place than in `goals.ts` where the users are?

Reply:
It's normally good practice to keep all the types in one place, for easy reference: think of it as protocol documentation.